apps/src/code-studio/hiddenStageRedux.js

```diff
@@ -2,48 +2,107 @@
  * Reducer and actions for stage lock info. This includes the teacher panel on
  * the course overview page, and the stage locking dialog.
  */
-
+import $ from 'jquery';
 import { INIT_PROGRESS } from './progressRedux';
 import experiments from '@cdo/apps/experiments';
 
-export const TOGGLE_HIDDEN = 'hiddenStage/TOGGLE_HIDDEN';
+export const UPDATE_HIDDEN_STAGES = 'hiddenStage/UPDATE_HIDDEN_STAGES';
+export const ALLOW_HIDEABLE = 'hiddenStage/ALLOW_HIDEABLE';
 
 export const hiddenStagesEnabled = () => experiments.isEnabled('hiddenStages');
 
+// State is a key indicating whether or not we've received hidden stage info
+// from the server, and then a set of stageIds mapped to bools that are true
+// if that stage is hidden
+const initialState = {
+  initialized: false
```

Comment:
Do we have a guarantee that no stage will be called `initialized`? If so, this might be fine.  If you do want to try Immutable.js I think you'd use an immutable `Hash` and call `mergeDeep`?